README.md
Rangeforce: developer-friendly range checks with user-friendly error messages
===============================================================================


Ever had to write the same old `if value < 0` bit of Python code to validate
a user input? Tired of unuseful error messages like _"Illegal value"_?
 
_Rangeforce_ is a very simple module offering multiple functions checking
the range of values, either integers or floats, or theirs lengths. It does
so in one line of code while providing an understandable error message that
can be displayed directly to the user as well.

```python
def classic_approach():
    value = int(input('Type a hex value: '), 16)
    if value < 0:
        raise ValueError('Value must be positive')
    elif value > 350:
        raise ValueError('Value must be <= 350')
    else:
        return value

# Same code, but simplified using Rangeforce
import rangeforce as rf

def with_rangeforce():
    value = int(input('Type a hex value: '), 16)
    return rf.limited(value, 0, 350)  # Here's the magic
```


Features
----------------------------------------


Installation
----------------------------------------


Example usage
----------------------------------------

```python
import rangeforce as rf

value = rf.limited(8000, 20, 5000, dtype=int)
# If successful, value will held 8000, otherwise (as it would happen in this
# example) raises a rangeforce.RangeError with a useful message:
# "Value must be in range [20, 5000]. 8000 found instead."
# Can be also shown directly to the user:

try:
    value = rf.limited(8000, 20, 5000, dtype=int)
except rf.RangeError as error:
    print(str(error))

# A missing bound (min or max) means unbounded
value = rf.limited(2000.0, None, 5000.0, dtype=float)
# Value must be <= 5000 but can be as small as it gets, including negative


# Especially useful in setters to validate the input in one line
class FullHdPicturesPixel(object):
    def __init__(self, x, y):
        self._x = None
        self._y = None
        self.x = x
        self.y = y
    
    @property
    def x(self):
        return self._x
    
    @x.setter
    def x(self, new_x):
        self._x = rf.limited(new_x, 0, 1920, 'The X pixel coordinate', int)
    
    @property
    def y(self):
        return self._y
    
    @y.setter
    def y(self, new_y):
        self._y = rf.limited(new_y, 0, 1080, 'The Y pixel coordinate', int)
        
pixel = FullHdPicturesPixel(10, 2000)
# This raises
# rangeforce.RangeError: The Y pixel coordinate must be in range [0, 1080]. 2000 found instead.


# Operating with C-like data structures and binary data?
# These functions might come in handy!
value = rf.int8(20) 
value = rf.int16(20) 
value = rf.int32(20) 
value = rf.int64(20)
value = rf.uint8(20) 
value = rf.uint16(20) 
value = rf.uint32(20) 
value = rf.uint64(20)

# Customize the name of the variable:
distance = rf.uint8(-3, 'Distance')
# This raises and error with the message:
# "Distance must be in range [0, 255]. -3 found instead."

# To check the length range of anything (e.g. a list or bytes):
value = rf.limited_len([1, 2, 3], 2, 7) # 'value' will hold [1, 2, 3]

# If you need an exact length, do it like this
pair = rf.exact_len([10, 20, 30], 2, name='pair of values')
# This raises and error with the message:
# "Length of pair of values must be exactly 2. 3 found instead.
```


CHANGELOG.md
Changelog
===============================================================================

All notable changes to this project will be documented in this file.

The format is based on 
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to 
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).


[1.0.0] - 2019-04-20
----------------------------------------

Initial version.


typeforce.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


class IllegalValueError(ValueError):
    pass


def clip(value, min, max):
    if value < min:
        return min
    elif value > max:
        return max
    else:
        return value


def limited_int(value, min, max, name='Value'):
    if not isinstance(value, int):
        raise TypeError(
            '{:} must be an integer. '
            '{:} found instead.'.format(name, type(value))
        )
    elif min is None and max is not None and value > max:
        raise IllegalValueError(
            '{:} must be in range ]-inf, {:}]. '
            '{:} found instead.'.format(name, max, value)
        )
    elif max is None and min is not None and value < min:
        raise IllegalValueError(
            '{:} must be in range [{:}, +inf[. '
            '{:} found instead.'.format(name, min, value)
        )
    elif min is not None and max is not None and (value < min or value > max):
        raise IllegalValueError(
            '{:} must be in range [{:}, {:}]. '
            '{:} found instead.'.format(name, min, max, value)
        )
    else:
        return value


def negative_int(value, name='Value'):
    return limited_int(value, None, -1, name)


def nonpositive_int(value, name='Value'):
    return limited_int(value, None, 0, name)


def positive_int(value, name='Value'):
    return limited_int(value, 1, None, name)


def nonnegative_int(value, name='Value'):
    return limited_int(value, 0, None, name)


def uint8(value, name='Value'):
    return limited_int(value, 0, 0xFF, name)


def uint16(value, name='Value'):
    return limited_int(value, 0, 0xFFFF, name)


def uint32(value, name='Value'):
    return limited_int(value, 0, 0xFFFFFFFF, name)


def uint64(value, name='Value'):
    return limited_int(value, 0, 0xFFFFFFFFFFFFFFFF, name)


def uint_bits(value, bits, name='Value'):
    return limited_int(value, 0, (1 << bits) - 1, name)


def int8(value, name='Value'):
    return limited_int(value, -0x80, 0x7F, name)


def int16(value, name='Value'):
    return limited_int(value, -0x8000, 0x7FFF, name)


def int32(value, name='Value'):
    return limited_int(value, -0x80000000, 0x7FFFFFFF, name)


def int64(value, name='Value'):
    return limited_int(value, -0x8000000000000000, 0x7FFFFFFFFFFFFFFF, name)


def limited_len(sized, min, max, name='Value', unit=''):
    length = len(sized)
    if min is None or min < 0:
        min = 0
    if max is not None:
        if max < 0:
            max = 0
        if length > max:
            raise IllegalValueError(
                'Length of {:} must be in range [{:}, {:}]{:}. '
                '{:} found instead.'.format(
                    name, min, max, unit or '' + unit, length)
            )
    else:
        return sized


def exact_len(sized, expected, name='Value', unit=''):
    length = len(sized)
    if expected is None or expected < 0:
        expected = 0
    if length != expected:
        raise IllegalValueError(
            'Length of {:} must be {:}{:}. '
            '{:} found instead.'.format(
                name, expected, unit or ' ' + unit, length)
        )


rangeforce.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright © 2019, Matjaž Guštin <dev@matjaz.it> <https://matjaz.it>.
# Released under the BSD 3-Clause License

"""Rangeforce, a developer-friendly check for "is this value
within the allowed range?" with user-friendly error messages.

Rangeforce aims to simplify the input validation process  where the error
message of the exception is shown directly to the user and has to be
understandable, which may happened in a command line interface or when using
a Python program interactively from a Python shell.

For example:

        value = int(input('How many hours per day do you sleep? '))
        value = rangeforce.limited(value, 0, 24, name='Hours of sleep')
        # Now value is valid. Otherwise an error message like this appears:
        # rangeforce.RangeError: Hours of sleep must be in range [0, 24]. 25 found instead.

        # Expecially useful for values that need to fit within an integer type:
        value = rangeforce.uint16(int(input('Type a 16-bit value: ')))
        rangeforce.RangeError: Value must be in range [0, 65535]. 70000 found instead.
"""

import math

__VERSION__ = 'v1.0.0'


class RangeError(Exception):
    """Value outside of the allowed range.

    A custom exception type raised by Rangeforce functions when the validation
    of the values fails, that is when the values are not within the acceptable
    bounds.
    """
    pass


def clip(value, min, max):
    """Clips (limits) the value to the given limits.

    The output is at least `min`, at most `max` and `value` if that value
    is between the `min` and `max`.

    Args:
        value: to be limited to [min, max]
        min: smallest acceptable value
        max: greatest acceptable value

    Returns:
        the given value if within [min, max] or min if value is smaller than
        min or max if value is greater than max.
    """
    if value < min:
        return min
    elif value > max:
        return max
    else:
        return value


def limited(value, min, max, name='Value', dtype=None):
    """Validates that value is within the [min, max] interval.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Either min or max can be set to None for an unbound validity interval, i.e.
    the value only has to be smaller or greater than something, not within a
    closed interval.

    The name of the value can be altered for a customized error message.

    The data type can be enforced if specified.

    Args:
        value: the value to be validated to be within [min, max]
        min: smallest acceptable value. Can be None if max is not None.
             Can be +inf, -inf. Cannot be NaN. Must be <= max.
        max: greatest acceptable value. Can be None if min is not None.
             Can be +inf, -inf. Cannot be NaN. Must be >= min.
        name: customizable name of the value that appears in the error message
        dtype: optional data type the value has to be

    Returns:
        the given value if within [min, max] and, optionally, of the correct
        data type

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not of the acceptable data type, if
                   specified.
        ValueError: if the min, max extremes are not valid (e.g. both None,
                    min greater than max, NaN etc.)

    Examples:
            >>> limited(0.5, 0, 1)  # Valid value
            0.5
            >>> limited(500, 0.1, 42)  # Value out of range
            rangeforce.RangeError: Value must be in range [0.1, -42]. 500
            found instead.
            >>> limited(50, 0, 24, name='Hours in a day')
            rangeforce.RangeError: Hours in a day must be in range [0,
            24]. 50 found instead.
            >>> limited(-1, 0, None, name='Earth satellites')
            rangeforce.RangeError: Earth satellites must be in range [0,
            +inf[. -1 found instead.
            >>> limited(1.1, 0, None, name='Earth satellites', dtype=int)
            TypeError: Earth satellites must be of type int. float found
            instead.
    """
    _validate_interval(min, max)
    _validate_type(name, value, dtype)
    if min is None and max is not None and (value > max or math.isnan(value)):
        raise RangeError(
            '{:} must be in range ]-inf, {:}]. '
            '{:} found instead.'.format(name, max, value)
        )
    elif max is None and min is not None and (
            value < min or math.isnan(value)):
        raise RangeError(
            '{:} must be in range [{:}, +inf[. '
            '{:} found instead.'.format(name, min, value)
        )
    elif min is not None and max is not None and (
            value < min or value > max or math.isnan(value)):
        raise RangeError(
            '{:} must be in range [{:}, {:}]. '
            '{:} found instead.'.format(name, min, max, value)
        )
    else:
        return value


def _validate_interval(min, max):
    if min is None and max is None:
        raise ValueError(
            '[min, max] interval must be closed on at least one extreme.')
    elif min is not None and math.isnan(min):
        raise ValueError('NaN is not a valid interval lower bound.')
    elif max is not None and math.isnan(max):
        raise ValueError('NaN is not a valid interval upper bound.')
    elif min is not None and max is not None and min > max:
        raise ValueError(
            'Interval extremes [{:}, {:}] not in order.'.format(min, max))


def _validate_type(name, value, dtype):
    if dtype is not None and not isinstance(value, dtype):
        raise TypeError(
            '{:} must be of type {:}. '
            '{:} found instead.'.format(name, dtype.__name__,
                                        type(value).__name__)
        )


def negative_int(value, name='Value'):
    """Validates that value is negative (< 0) and of type int.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within ]-inf, 0[
        name: customizable name of the value that appears in the error message

    Returns:
        the given value if < 0

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, None, -1, name, dtype=int)


def nonpositive_int(value, name='Value'):
    """Validates that value is non-positive (<= 0) and of type int.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within ]-inf, 0]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value if <= 0

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, None, 0, name, dtype=int)


def positive_int(value, name='Value'):
    """Validates that value is positive (> 0) and of type int.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within ]0, +inf[
        name: customizable name of the value that appears in the error message

    Returns:
        the given value if > 0

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 1, None, name, dtype=int)


def nonnegative_int(value, name='Value'):
    """Validates that value is non-negative (>= 0) and of type int.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [0, +inf[
        name: customizable name of the value that appears in the error message

    Returns:
        the given value if >= 0

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 0, None, name, dtype=int)


def uint8(value, name='Value'):
    """Validates that value fits in an 8-bit unsigned integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [0, 255]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 8 bits (unsigned)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 0, 0xFF, name, dtype=int)


def uint16(value, name='Value'):
    """Validates that value fits in a 16-bit unsigned integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [0, 65535]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 16 bits (unsigned)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 0, 0xFFFF, name, dtype=int)


def uint32(value, name='Value'):
    """Validates that value fits in a 32-bit unsigned integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [0, 4294967295]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 32 bits (unsigned)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 0, 0xFFFFFFFF, name, dtype=int)


def uint64(value, name='Value'):
    """Validates that value fits in a 64-bit unsigned integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [0, 18446744073709551615]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 64 bits (unsigned)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 0, 0xFFFFFFFFFFFFFFFF, name, dtype=int)


def uint_bits(value, bits, name='Value'):
    """Validates that value fits in an unsigned integer of specified bitlength.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [0, 2**bits-1]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in the given amount of  bits
        (unsigned)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, 0, (1 << bits) - 1, name, dtype=int)


def int8(value, name='Value'):
    """Validates that value fits in an 8-bit signed integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [-128, 127]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 8 bits (signed)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, -0x80, 0x7F, name, dtype=int)


def int16(value, name='Value'):
    """Validates that value fits in a 16-bit signed integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [-32768, 32767]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 16 bits (signed)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, -0x8000, 0x7FFF, name, dtype=int)


def int32(value, name='Value'):
    """Validates that value fits in a 32-bit signed integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [-2147483648, 2147483647]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 32 bits (signed)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, -0x80000000, 0x7FFFFFFF, name, dtype=int)


def int64(value, name='Value'):
    """Validates that value fits in a 16-bit signed integer.

    If the value is valid, it returns the value itself.
    If the value is not valid, it raises a RangeError with an understandable
    error message that includes expected range and failing value.

    Args:
        value: the value to be validated to be within [-9223372036854775808,
               9223372036854775807]
        name: customizable name of the value that appears in the error message

    Returns:
        the given value can be expressed in 64 bits (signed)

    Raises:
        RangeError: if the value is not within the acceptable range.
        TypeError: if the value is not an integer.
    """
    return limited(value, -0x8000000000000000, 0x7FFFFFFFFFFFFFFF, name,
                   dtype=int)


def limited_len(sized, min, max, name='value'):
    """Validates that value has a length within the [min, max] interval.

    If the sized value is valid, it returns the value itself.
    If the sized value is not valid, it raises a RangeError with an
    understandable error message that includes expected length range and
    failing sized value.

    Either min or max can be set to None for an unbound validity
    interval, i.e. the value only has to have a length smaller or greater than
    something, not within a closed interval.

    The name of the sized value can be altered for a customized error message.

    Args:
        sized: the value whose length is to be validated to be within
               [min, max]
        min: smallest acceptable length. Can be None if max is not None.
             Can be +inf, -inf. Cannot be NaN. Must be <= max and >= 0.
        max: greatest acceptable length. Can be None if min is not None.
             Can be +inf, -inf. Cannot be NaN. Must be >= min and >= 0.
        name: customizable name of the value that appears in the error message

    Returns:
        the given sized value if has length within [min, max]

    Raises:
        RangeError: if the value does not have a length within the acceptable
                    range.
        ValueError: if the min, max extremes are not valid (e.g. negative,
                    both None, min greater than max, NaN etc.)

    Examples:
            >>> limited_len([1, 2, 3], 0, 10)  # Valid value
            [1, 2, 3]
            >>> limited_len([1, 2, 3], 0, 2)  # Value out of range
            rangeforce.RangeError: Length of value must be in range [0,
            2]. 3 found instead.
            >>> limited_len([1, 2, 3, 4], 0, 3, name='groups')
            rangeforce.RangeError: Length of groups must be in range [0,
            3]. 4 found instead.
            >>> limited_len([1, 2, 3], 10, None)
            rangeforce.RangeError: Length of value must be in range [10,
            +inf[. 3 found instead.
    """
    length = len(sized)
    _validate_non_negative_interval_extremes(min, max)
    limited(length, min, max, name='Length of ' + name)
    return sized


def _validate_non_negative_interval_extremes(min, max):
    if min is not None and min < 0:
        raise ValueError(
            'Length lower bound must be non-negative. '
            '{:} found instead.'.format(min)
        )
    elif max is not None and max < 0:
        raise ValueError(
            'Length upper bound must be non-negative. '
            '{:} found instead.'.format(max)
        )


def exact_len(sized, expected, name='value'):
    """Validates that value has an exact length.

    If the sized value is valid, it returns the value itself.
    If the sized value is not valid, it raises a RangeError with an
    understandable error message that includes expected length and
    failing sized value.

    The name of the sized value can be altered for a customized error message.

    Args:
        sized: the value whose length is to be validated to be exactly as
               expected.
        expected: only acceptable length. Must be an integer >= 0.
        name: customizable name of the value that appears in the error message

    Returns:
        the given sized value if has length matching the expected

    Raises:
        RangeError: if the value does not have a length matching the expected.
        ValueError: if the expected length is not valid (e.g. not integer,
                    negative, None)

    Examples:
            >>> exact_len([1, 2, 3], 3)  # Valid value
            [1, 2, 3]
            >>> exact_len([1, 2, 3], 2)
            rangeforce.RangeError: Length of value must be exactly 2. 3
            found instead.
            >>> exact_len([1], 2, name='pairs')
            rangeforce.RangeError: Length of pairs must be exactly 2. 1
            found instead.
    """
    length = len(sized)
    _validate_expected_length(expected)
    if length != expected:
        raise RangeError(
            'Length of {:} must be exactly {:}. '
            '{:} found instead.'.format(name, expected, length)
        )
    return sized


def _validate_expected_length(expected):
    if not isinstance(expected, int):
        raise TypeError(
            'Expected length must be an integer. '
            '{:} found instead.'.format(type(expected).__name__)
        )
    elif expected < 0:
        raise ValueError(
            'Expected length must be non-negative. '
            '{:} found instead.'.format(expected)
        )
